### PR TITLE
Zotero api route

### PR DESCRIPTION
**Pull request by skybristol**
This pull request contains three notebooks that work through a different process to assemble ITIS scientific names that operates through the Zotero API with direct access to the online library. This addresses the issue that I posted on this [topic](https://github.com/usgs-bcb/wlci-species/issues/6). Essentially, if we want project scientists to maintain their reference library via Zotero and share it via Zotero cloud, then it will be more sustainable and scalable if we have some way of keeping all of the fundamental value-added information we have software go out and gather together in the Zotero library.

I was pleasantly surprised at how reasonable it was to interact with the Zotero API to make this work. The pyzotero package I used is pretty comprehensive, although I did have to tweak on one of the functions and sent the project a [pull request](https://github.com/urschrei/pyzotero/pull/105) with that change (so install my branch if you want to check this out for yourself). The attachments idea in the Zotero information model is pretty reasonable for accommodating the links to xDD derivative/digested documents legitimately as link attachments on metadata items in the Zotero library that could be added by both software and human users. The same attachment idea can also legitimately hold the results of hitting the API and returning content from one of the dictionary indexing processes, and this also can add immediate value to library users.

The cool thing about all of this is that we end up with a live, multi-modal way of interacting with the reference library for a given project. Users can go back and forth between the desktop client and the web client, with the same information and functionality available in both. We can write code against the API to access and work with the library for downstream information gathering processes. The way I wrote the last function to pull scientific names back from ITIS, that should work for any dictionary that xDD sets up against documents.

The only real issue with this architecture is how it ultimately gets operationalized. You'll see in my notebooks, that I created an API key with permissions specific to my user account and the WLCI library. That's good security and how we'd want the system to operate, but it means that any user or process that's going to run this workflow will need to have an API key with the appropriate permissions. We could conceivably set up some cyberinfrastructure that checks a registered list of Zotero libraries over time for updates, runs a lambda function to establish linkages to xDD, and then runs lambda functions to update or add any new dictionary terms. From an Information System (BIS or something broader) perspective, a part of the Research Reference Registry/Library then would consist of registered shared libraries that drive that whole process. Presumably, we could instantiate the same basic idea using any of these types of systems that have some kind of online presence with an API that we can operate against and an information model that accommodates the type of information we are putting together here. The cool thing about Zotero, though, is that read access to public content from libraries can be accessed without an API key. We need that for writing content back to the library, but we can take advantage of the information content in other ways without.

The one big potential hiccup for us on this stuff is whatever is going on with the Department and USGS IT staff cracking down on any use of "cloud" services that aren't on some kind of accepted list with an "Authority to Operate." What we are doing here is all based on completely public information that is already on the web and accessible, so we're not running into any need for FSP review and approval for the information content itself. However, it is making use of a free tier cloud-based information storage solution provided by a vendor that would likely put it into the same category with other things the IT folks seem to be worried about. If that comes up as a problem, then our only real option is to go back to some other solution based on established internal capabilities like ScienceBase. However, from a usability standpoint, that would be a huge loss in terms of what end users experience and what they can do. Even though we could use the ScienceBase API to do much of what I did here in this contribution, users would lose out on all of the powerful functionality that is nicely built into Zotero (or Mendeley, EndNote, etc.):

* Ease of adding references to the library from clicking a tool in a web browser
* All of the specialized reference management stuff built into the tools (e.g., reformatting for different types of citation/references in pubs, deduplication, etc.)
* Any type of desktop client that facilitates more robust library management and collaboration
* Plugins for manuscript development tools (Google Docs, Word, etc.)

Please review these ideas and run the code for yourselves to see how it operates. The "attachments" are already built in the WLCI Zotero library, so you can skip ahead to the notebook that simply pulls together the species list from that route, but check out how the process worked and provide comments and input on how to make it better.

Note: I also added an environment.yml to the root of the project in one of the commits for this branch. I think I hit all the high level requirements, but change it up if I missed anything. I'm still learning that process and how best to build those. I have found that you really only want to include the base packages needed and let them deal with any internal dependencies. You can then create a conda environment for your work with...

``conda env create -f environment.yml``

...for setting up an environment to work with any given project.

### CODE REVIEW

File: environment.yml [hunk 1]
```diff
@@ -10,9 +10,9 @@ dependencies:
   - pandas
   - pip
   - requests
-  - pyzotero
   - pip:
     - sciencebasepy
     - git+git://github.com/usgs-bcb/bispy.git
+    - git+https://github.com/skybristol/pyzotero@attachment-template-type#egg=pyzotero
```
Comment (dwief-usgs) [inline environment.yml]: in creating a new environment from this yml file, none of the packages indented under pip install (sciencebasepy, bispy, pyzotero)

File: workflow/Pull References via Zotero API.ipynb [hunk 2]
```diff
@@ -0,0 +1,689 @@
...
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": 7,
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "lookup_doi_list = [(i[\"data\"][\"key\"],i[\"data\"][\"DOI\"]) for i in wlci_lib_items if \"DOI\" in i[\"data\"].keys() and len(i[\"data\"][\"DOI\"]) > 0]\n",
+    "lookup_doi_list.extend([(i[\"data\"][\"key\"],i[\"data\"][\"extra\"]) for i in wlci_lib_items if \"DOI\" not in i[\"data\"].keys() and len(i[\"data\"][\"extra\"]) > 0 and i[\"data\"][\"extra\"].split(\":\")[0] != \"OCLC\"])"
```
Comment (dwief-usgs) [inline workflow/Pull References via Zotero API.ipynb]: lookup_doi_list.extend([(i["data"]["key"],i["data"]["extra"]) for i in wlci_lib_items if "DOI" not in i["data"].keys() and 'extra' in i['data'].keys() and len(i["data"]["extra"]) > 0 and i["data"]["extra"].split(":")[0] != "OCLC"] )

needed to check to see if extra key existed... must not exist for everything as it threw a keyerror

File: environment.yml (hunk 1 above)
Comment (skybristol) [reply environment.yml]: How are you creating the environment? Did you see any errors in the env creation process?

I just went in, blew away my old wlci-species Conda env, and then created a new one with

``conda env create -f environment.yml``

Running...

``conda list``

...afterward shows all of the packages installed by conda and pip.

Comment (dwief-usgs) [reply environment.yml]: Correct, I used the same methods (as posted above) to create the env.  Although it appears to be picking up all packages on the creation of the env those with pip subprocess are not represented in the env as shown in the conda list.  

*image showing create env
Capture

*image showing sciencebasepy and pyzotero missing
Capture2

**Comment by dwief-usgs** [comment]
@skybristol  any reason to leave this branch open?

**Comment by skybristol** [comment]
Might as well close out the branch since you merged. We can open up something new for follow on work.

File: environment.yml (hunk 1 above)
Comment (gcdonovan) [reply environment.yml]: Everything went through ok for me...
Screen Shot 2019-10-01 at 4 11 04 PM

Comment (dwief-usgs) [reply environment.yml]: @gcdonovan you were supposed to say it must have been a windows issue ;)

glad it worked for you

**Pull request merged**